### PR TITLE
wip: call fcu on invalid payload

### PR DESCRIPTION
**Pull request by terencechain**
After fcu, the node should update its head and attempt to call fcu again with the latest head

### CODE REVIEW

File: beacon-chain/blockchain/execution_engine.go [hunk 1]
```diff
@@ -90,6 +90,33 @@ func (s *Service) notifyForkchoiceUpdate(ctx context.Context, headState state.Be
...
+			b, err := s.cfg.BeaconDB.Block(ctx, r)
+			if err != nil {
+				return nil, err
+			}
+			st, err := s.cfg.StateGen.StateByRoot(ctx, r)
+			if err != nil {
+				return nil, err
+			}
+			fRoot := st.FinalizedCheckpoint().Root
+			_, err = s.notifyForkchoiceUpdate(ctx, st, b.Block(), r, bytesutil.ToBytes32(fRoot))
```
Comment (potuz) [inline beacon-chain/blockchain/execution_engine.go:116]: This became a recursive function, we should check for context cancellation.

File: beacon-chain/blockchain/execution_engine.go [hunk 2]
```diff
@@ -90,6 +90,33 @@ func (s *Service) notifyForkchoiceUpdate(ctx context.Context, headState state.Be
...
+			}
+			b, err := s.cfg.BeaconDB.Block(ctx, r)
+			if err != nil {
+				return nil, err
+			}
+			st, err := s.cfg.StateGen.StateByRoot(ctx, r)
+			if err != nil {
+				return nil, err
+			}
+			fRoot := st.FinalizedCheckpoint().Root
```
Comment (potuz) [inline beacon-chain/blockchain/execution_engine.go:115]: Shouldn't we send the current `finalizedRoot`? if we used `st.FinalizedCheckpoint().Root` and the INVALID chain crossed an epoch boundary we will revert finalization.

File: beacon-chain/blockchain/execution_engine_test.go [hunk 3]
```diff
@@ -191,6 +196,49 @@ func Test_NotifyForkchoiceUpdate(t *testing.T) {
 	}
 }
 
+func Test_NotifyForkchoiceUpdateRecursive(t *testing.T) {
```
Comment (rauljordan) [inline beacon-chain/blockchain/execution_engine_test.go:199]: Can we add more useful checks that will actually test for the behavior we want? For example, we want to ensure that forkchoice is updated with a correct head after fork choice selection upon first receiving an invalid payload from EL. If we can simulate a fork where one branch contains an invalid payload and another is valid, we should prove that forkchoice was updated to the better head, and that forkchoice updated was called with the right information

File: beacon-chain/forkchoice/doubly-linked-tree/optimistic_sync.go [hunk 4]
```diff
@@ -81,7 +81,7 @@ func (s *Store) removeNode(ctx context.Context, node *Node) ([][32]byte, error)
 				if i != len(children)-1 {
 					children[i] = children[len(children)-1]
 				}
-				node.parent.children = children[:len(children)-2]
+				node.parent.children = children[:len(children)-1]
```
Comment (potuz) [inline beacon-chain/forkchoice/doubly-linked-tree/optimistic_sync.go:84]: This was an off-by one in forkchoice, need to add an extra test for this off by one, will add it as a PR to forkchoice

**Review by rauljordan — APPROVED** [review]
This test case is awesome - thank you for putting this together
